stardist/bioimageio_utils.py

```diff
@@ -0,0 +1,426 @@
+    assert n_outputs == 1
+    out_paths = [outdir / "test_output.npy"]
+    np.save(out_paths[0], test_outputs)
+
+    from tensorflow import __version__ as tf_version
+    data = dict(weight_uri=assets_uri, test_inputs=[in_path], test_outputs=out_paths,
+                config=config, tensorflow_version=tf_version)
+    data.update(input_config)
+    data.update(output_config)
+    if is_2D:
+        data.update(attachments=dict(files=[str(macro_file)]))
+
+    return data
+
+
+def export_bioimageio(
+    model,
+    outpath,
+    test_input,
+    test_input_axes=None,
+    test_input_norm_axes='ZYX',
+    name="bioimageio_model",
+    mode="tensorflow_saved_model_bundle",
+    min_percentile=1.0,
+    max_percentile=99.8,
+    overwrite_spec_kwargs={}
+):
+    """Export stardist model into bioimageio format, https://github.com/bioimage-io/spec-bioimage-io.
+
+    Parameters
+    ----------
+    model: StarDist2D, StarDist3D
+        the model to convert
+    outpath: str, Path
+        where to save the model
+    test_input: np.ndarray
+        input image for generating test data
+    test_input_axes: str or None
+         the axes of the test input, for example 'YX' for a 2d image or 'ZYX' for a 3d volume
+         using None assumes that axes of test_input are the same as those of model
+    test_input_norm_axes: str
+         the axes of the test input which will be jointly normalized, for example 'ZYX' for all spatial dimensions ('Z' ignored for 2D input)
+         use 'ZYXC' to also jointly normalize channels (e.g. for RGB input images)
+    name: str
+        the name of this model (default: "StarDist Model")
+    mode: str
+        the export type for this model (default: "tensorflow_saved_model_bundle")
+    overwrite_spec_kwargs: dict
+        spec keywords that should be overloaded (default: {})
+    """
+    _, build_model, *_ = _import()
+    from stardist.models import StarDist2D, StarDist3D
+    isinstance(model, (StarDist2D, StarDist3D)) or _raise(ValueError("not a valid model"))
+    0 <= min_percentile < max_percentile <= 100 or _raise(ValueError("invalid percentile values"))
+
+    outpath = Path(outpath)
+    if outpath.suffix == "":
+        outdir = outpath
+        zip_path = outdir / f"{name}.zip"
+    elif outpath.suffix == ".zip":
+        outdir = outpath.parent
+        zip_path = outpath
+    else:
+        raise ValueError(f"outpath has to be a folder or zip file, got {outpath}")
+    outdir.mkdir(exist_ok=True, parents=True)
+
+    kwargs = _get_stardist_metadata(outdir)
+    model_kwargs = _get_weights_and_model_metadata(outdir, model, test_input, test_input_axes, test_input_norm_axes, mode,
+                                                   min_percentile=min_percentile, max_percentile=max_percentile)
+    kwargs.update(model_kwargs)
+    kwargs.update(overwrite_spec_kwargs)
+
+    build_model(name=name, output_path=zip_path, add_deepimagej_config=(model.config.n_dim==2), **kwargs)
```

Comment:
`build_model` is very messy in the sense that it is spewing files everywhere. How can that be stopped (i.e. use a temp folder instead)? All we want is the zip at the end, or not?

The parameter `output_path` seems to have been removed?

Reply:
These are the generated files by calling `export_bioimageio(model, Path('./out'), ...)`:

```
$ tree
.
├── out
│   ├── 2D_versatile_fluo.zip
│   ├── README.md
│   ├── requirements.txt
│   ├── stardist_postprocessing.ijm
│   ├── stardist_weights.h5
│   ├── test_input.npy
│   ├── test_output.npy
│   ├── tf_model
│   │   ├── saved_model.pb
│   │   └── variables
│   │       ├── variables.data-00000-of-00001
│   │       └── variables.index
│   └── TF_SavedModel.zip
├── per_sample_scale_range.ijm
├── README.md
├── sample_input_0.tif
├── sample_output_0.tif
├── stardist_logo.jpg
├── stardist_postprocessing.ijm
├── stardist_weights.h5
├── test_input.npy
├── test_output.npy
└── TF_SavedModel.zip
```

Reply:
Hmm, somehow it doesn't use the correct tmp path. I will double check.

Reply:
@uschmidt83 ok, it turns out that I forgot to specify a tmp path.
I have proposed a solution in https://github.com/constantinpape/stardist/pull/1.
Note that this only works if the tmp_dir is **not** below the cwd. This is due to how path resolution in the bioimageio library works and fixing this will be a bit tricky.

I have just proposed a very simple solution without any cleanup of the tmp_dir, feel free to just merge this or expand the solution and add some clean up.